### PR TITLE
[#155426] Lockout password-based user after 5 failed login attempts

### PR DESCRIPTION
**Pull request by wangminghao**
https://pm.tablexi.com/issues/155426

- Devise set up for locking accounts with custom error message
- Set User as lockable
- Create LogEvent when User becomes locked

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -197,4 +197,9 @@ def create_default_price_group!
     price_group_members.find_or_create_by!(price_group: default_price_group)
   end
 
+  def lock_access!(opts = { })
+    super
+    LogEvent.log(self, :access_locked, self, metadata: {cause: "Too many login attempts"})
+    puts "\n========= HELLO"
```
Comment (rdunlop) [inline app/models/user.rb:203]: probably don't need this

**Review by rdunlop — COMMENTED** [review]
I seem to recall discussion with the client that "doing a password reset should clear the locked-out state" so that if they did a reset-password, they could log in immediately.

Is this possible? desired?

**Comment by wangminghao** [comment]
> I seem to recall discussion with the client that "doing a password reset should clear the locked-out state" so that if they did a reset-password, they could log in immediately.
> 
> Is this possible? desired?

My impression was that the desired behavior was the "normal" devise behavior, but I don't have the client context. @giladshanan do you have insight into this?

**Comment by giladshanan** [comment]
The goal here is to improve security with minimal effort and impact on users.  We have a couple of options here:
- the user waits 5 min
- use the `:email` or `:both` unlock strategy, which are unlocked by password reset (https://github.com/heartcombo/devise/issues/2090) 

I will check in with the client this afternoon.

**Comment by giladshanan** [comment]
The client is in favor of the `:email` unlock strategy instead of the 5 min cool off.

**Review by giladshanan — CHANGES_REQUESTED** [review]
One small request, looks good!

**Review by rdunlop — APPROVED** [review]

**Review by rdunlop — APPROVED** [review]

**Review by giladshanan — APPROVED** [review]

**Pull request merged**